scipy/fft/tests/test_basic.py

```diff
@@ -19,240 +32,358 @@ def fft1(x):
+            fft.ihfft2(x, norm="ortho")
+        )
+        assert_close(expect * (30 * 20), fft.ihfft2(x, norm="forward"))
+
+    @skip_if_array_api_gpu
+    @array_api_compatible
+    def test_hfftn(self, xp):
+        x = xp.asarray(random((30, 20, 10)))
+        assert_close(x, fft.hfftn(fft.ihfftn(x)))
         for norm in ["backward", "ortho", "forward"]:
-            assert_array_almost_equal(
-                x, fft.hfftn(fft.ihfftn(x, norm=norm), norm=norm))
+            assert_close(x, fft.hfftn(fft.ihfftn(x, norm=norm), norm=norm))
 
-    def test_ihfftn(self):
-        x = random((30, 20, 10))
+    @skip_if_array_api_gpu
+    @array_api_compatible
+    @skip_if_array_api_backend('torch')
+    def test_ihfftn(self, xp):
+        x = xp.asarray(random((30, 20, 10)))
         expect = fft.ifftn(x)[:, :, :6]
-        assert_array_almost_equal(expect, fft.ihfftn(x))
-        assert_array_almost_equal(expect, fft.ihfftn(x, norm="backward"))
-        assert_array_almost_equal(expect * np.sqrt(30 * 20 * 10),
-                                  fft.ihfftn(x, norm="ortho"))
-        assert_array_almost_equal(expect * (30 * 20 * 10),
-                                  fft.ihfftn(x, norm="forward"))
+        assert_close(expect, fft.ihfftn(x))
+        assert_close(expect, fft.ihfftn(x, norm="backward"))
+        assert_close(
+            expect * xp.sqrt(xp.asarray(30 * 20 * 10, dtype=xp.float64)),
+            fft.ihfftn(x, norm="ortho")
+        )
+        assert_close(expect * (30 * 20 * 10), fft.ihfftn(x, norm="forward"))
 
+    # torch.fft not yet implemented by array-api-compat
+    @skip_if_array_api_backend('torch')
+    @array_api_compatible
     @pytest.mark.parametrize("op", [fft.fftn, fft.ifftn,
-                                    fft.rfftn, fft.irfftn,
-                                    fft.hfftn, fft.ihfftn])
-    def test_axes(self, op):
-        x = random((30, 20, 10))
-        axes = [(0, 1, 2), (0, 2, 1), (1, 0, 2), (1, 2, 0), (2, 0, 1), (2, 1, 0)]
+                                    fft.rfftn, fft.irfftn])
+    def test_axes_standard(self, op, xp):
+        x = xp.asarray(random((30, 20, 10)))
+        axes = [(0, 1, 2), (0, 2, 1), (1, 0, 2),
+                (1, 2, 0), (2, 0, 1), (2, 1, 0)]
+        xp_test = array_namespace(x)
         for a in axes:
-            op_tr = op(np.transpose(x, a))
-            tr_op = np.transpose(op(x, axes=a), a)
-            assert_array_almost_equal(op_tr, tr_op)
+            op_tr = op(xp_test.permute_dims(x, axes=a))
+            tr_op = xp_test.permute_dims(op(x, axes=a), axes=a)
+            assert_close(op_tr, tr_op)
+
+    @skip_if_array_api_gpu
+    @array_api_compatible
+    @pytest.mark.parametrize("op", [fft.hfftn, fft.ihfftn])
+    def test_axes_non_standard(self, op, xp):
+        x = xp.asarray(random((30, 20, 10)))
+        axes = [(0, 1, 2), (0, 2, 1), (1, 0, 2),
+                (1, 2, 0), (2, 0, 1), (2, 1, 0)]
+        xp_test = array_namespace(x)
```

Comment:
I was staring at this code for quite a bit to figure out why there is so much more test code now. The test logic itself is duplicated, which is problematic. I could think of two better ways of writing these, to avoid this - see https://github.com/rgommers/scipy/commit/2ab420106d40cb57ff57e53f7dac2e5ca4596571 (not 100% working, but I hope it's clear). That already contains more removals than additions, and it will make the actual diff in this PR even smaller because there's a couple of changes that are straight undo's of changes in this PR.

Reply:
Addressed https://github.com/scipy/scipy/pull/19005/commits/5ff56152c5ed2a57b9fee653e78d08051af426e8.

Reply:
LGTM now, thanks.